lib/std/Thread/RecursiveMutex.zig

```diff
@@ -0,0 +1,280 @@
+pub fn lock(self: *RecursiveMutex) void {
+    self.impl.lock();
+}
+
+/// Releases the mutex which was previously acquired with `lock()` or `tryLock()`.
+/// It is undefined behavior if the mutex is unlocked from a different thread that it was locked from.
+pub fn unlock(self: *RecursiveMutex) void {
+    self.impl.unlock();
+}
+
+const SingleThreadedImpl = struct {
+    is_locked: bool = false,
+    recursion_depth: usize = 0,
+
+    fn tryLock(self: *Impl) bool {
+        self.recursion_depth += 1;
+        return true;
+    }
+
+    fn lock(self: *Impl) void {
+        if (self.recursion_depth == 0) {
+            self.is_locked = true;
+        }
+        self.recursion_depth += 1;
+    }
+
+    fn unlock(self: *Impl) void {
+        assert(self.is_locked);
+        const depth =
+            self.recursion_depth - 1;
+        if (depth == 0) {
+            self.is_locked = false;
+        }
+        self.recursion_depth = depth;
+    }
+};
+
+const FutexImpl = struct {
+    // Technically, there are only 2 states, but we use a whole u32 here
+    // because this value is awaited by a Futex
+    const State = enum(u32) { unlocked, locked };
+    state: Atomic(u32) =
+        Atomic(u32).init(0),
+    owning_thread_id: Atomic(Thread.Id) =
+        Atomic(Thread.Id).init(0),
+    recursion_depth: Atomic(usize) =
+        Atomic(usize).init(0),
+
+    pub fn tryLock(self: *Impl) bool {
+        const current_thread_id =
+            Thread.getCurrentId();
+        // != null <=> was locked before
+        if (self.state.compareAndSwap(@enumToInt(State.unlocked), @enumToInt(State.locked), .AcqRel, .Acquire) != null) {
+            const owning_thread = self.owning_thread_id.load(.Acquire);
+            if (owning_thread != current_thread_id) {
+                return false;
+            }
+        }
+        // record thread id & break early: no need for blocking
+        self.owning_thread_id.store(current_thread_id, .Release);
+        _ =
+            self.recursion_depth.fetchAdd(1, .Monotonic);
+        return true;
+    }
+
+    pub fn lock(self: *Impl) void {
+        const current_thread_id =
+            Thread.getCurrentId();
+        var state_result: ?u32 = undefined;
+        while (true) {
+            state_result =
+                self.state.compareAndSwap(@enumToInt(State.unlocked), @enumToInt(State.locked), .AcqRel, .Acquire);
+            // was unlocked & this thread made it locked
+            if (state_result == null) {
+                // sanity check
+                assert(self.recursion_depth.value == 0);
+                // record thread id & break early: no need for blocking
+                self.owning_thread_id.store(current_thread_id, .Release);
+                break;
+            } else {
+                const owning_thread =
+                    self.owning_thread_id.load(.Acquire);
+                if (owning_thread == current_thread_id) {
+                    break;
+                }
```

Comment:
I think this is a race and will definitely be buggy if `Thread.getCurrentId()` ever returns 0.

To demonstrate lets have thread's 0 and 1, try locking the mutex at the same time with the following scheduling:
1. the mutex is unlocked, (i.e. `self.state == unlocked`)
2. T1 successfully completes the compare exchange T0 fails -> `self.state == locked`
3. T0 executes the else branch and sees that `owning_thread == 0`, and starts executing the critical section
4. T1 changes `owning_thread = 1` and also executes the critical section

If the "unlocked" value of `self.owning_thread` would be an invalid thread identifier, then this couldn't happen, but I'm still not totally confident there are no other lurking problems, it would be best if the locked state and the owning thread could be modified by a single atomic operation / cas loop.

Reply:
Thanks for taking a look at this PR!

>  will definitely be buggy if Thread.getCurrentId() ever returns 0.

As far as I understand, this should "never happen". In the Windows case,[ official docs say so explicitly](https://learn.microsoft.com/en-us/windows/win32/procthread/thread-handles-and-identifiers?redirectedfrom=MSDN). For other targets, I found other code on internet which used `0` as an "invalid value2 sentinel, but the man pages do not explicitly mention this, so I'm not sure what to think...

> it would be best if the locked state and the owning thread could be modified by a single atomic operation / cas loop.

I agree! Furthermore, I think it's awkward that there are 2 variables expressing the state (e.g. `owner == 0 ` and `state == .locked` are expressing the same state).

If `thread_id ` can truly be any value within the u64 range (including 0), then there is no way to distinguish whether or not the value is set (that extra bit of information must be stored somewhere else).
So, there must be 2 variables (1 for state, and one for the owning address). Both need to be modified atomically "at the same time", which so I suppose for this we should use a different implementation.
Some ideas are:
- use `cmpxchg16b `to compare the 128 bits (not sure how widely supported this instruction is, though)
- abandon Atomics + Futex in favour of mutex + condition variable

If 0 is truly an invalid value, then we can get rid of  state, and use the `Futex `to only look at `owner_thread_id`.
However, that would require extending Futex to be able to wait on 64-bit values. I imagine this to be useful for other places as well.

From a cursory search, at least Windows and Darwin native APIs seem to support this, so perhaps this is worth looking into.
https://learn.microsoft.com/en-us/windows/win32/api/synchapi/nf-synchapi-waitonaddress
https://github.com/apple/darwin-xnu/blob/2ff845c2e033bd0ff64b5b6aa6063a1f8f65aa32/bsd/kern/sys_ulock.c#L513

I think in this current state the PR is not in a state where it can be merged, because we first need to address the issues you kindly pointed out.
However, I'd like to hear more opinions on whether or not Recursive Mutex should be part of the standard library in the first place before tackling the problems above.